clap_builder/src/builder/arg.rs

```diff
@@ -4364,7 +4364,11 @@ impl Arg {
 
         debug_assert!(self.is_takes_value_set());
         for (n, val_name) in val_names.iter().enumerate() {
-            let arg_name = if self.is_positional() && (num_vals.min_values() == 0 || !required) {
+            let all_optional = self.get_min_vals() == 0;
+            let arg_name = if match self.is_positional() {
+                true => num_vals.min_values() == 0 || !required,
+                false => !all_optional && (n + 1 > num_vals.min_values()),
+            } {
```

Comment:
imo this is too complicated if an `if` expression.

One option for simplifying this is to calculate an index before the loop for when optional values start

Reply:
I've moved the check out into a closure, simplified the logic, and added some explanation to the non-positional case, which hopefully makes things a bit clearer.